source/client-side-encryption/client-side-encryption.md

```diff
@@ -0,0 +1,2483 @@
+provider types. Additionally, they MUST document that masterKey is **required** for these KMS provider types and is not
+optional.
+
+The value of `endpoint` or `keyVaultEndpoint` is a host name with optional port number separated by a colon. E.g.
+"kms.us-east-1.amazonaws.com" or "kms.us-east-1.amazonaws.com:443". It is assumed that the host name is not an IP
+address or IP literal. Though drivers MUST NOT inspect the value of "endpoint" that a user sets when creating a data
+key, a driver will inspect it when connecting to KMS to determine a port number if present.
+
+#### keyAltNames
+
+An optional list of string alternate names used to reference a key. If a key is created with alternate names, then
+encryption may refer to the key by the unique alternate name instead of by `_id`. The following example shows creating
+and referring to a data key by alternate name:
+
+```python
+opts = DataKeyOpts(keyAltNames=["name1"])
+clientencryption.create_data_key ("local", opts)
+# reference the key with the alternate name
+opts = EncryptOpts(keyAltName="name1", algorithm="AEAD_AES_256_CBC_HMAC_SHA_512-Random")
+clientencryption.encrypt("457-55-5462", opts)
+```
+
+#### keyMaterial
+
+An optional BinData of 96 bytes to use as custom key material for the data key being created. If `keyMaterial` is given,
+the custom key material is used for encrypting and decrypting data. Otherwise, the key material for the new data key is
+generated from a cryptographically secure random device.
+
+### rewrapManyDataKey
+
+If applicable, drivers MUST document that users must upgrade dependencies if necessary to avoid being impacted by
+[MONGOCRYPT-464](https://jira.mongodb.org/browse/MONGOCRYPT-464).
+
+If applicable, drivers MUST return an error if rewrapManyDataKey is called with libmongocrypt 1.5.1 or 1.5.0.
+
+### RewrapManyDataKeyOpts
+
+```typescript
+class RewrapManyDataKeyOpts {
+   provider: String
+   masterKey: Optional<Document>
+}
+```
+
+The `masterKey` document MUST have the fields corresponding to the given `provider` as specified in
+[masterKey](#masterkey). `masterKey` MUST NOT be given if it is not applicable for the given `provider`.
+
+### RewrapManyDataKeyResult
+
+```typescript
+class RewrapManyDataKeyResult {
+   bulkWriteResult: Optional<BulkWriteResult>;
+}
+```
+
+`bulkWriteResult` is the [result of the bulk write operation](../crud/crud.md##write-results) used to update the key
+vault collection with one or more rewrapped data keys. If `rewrapManyDataKey()` does not find any matching keys to
+rewrap, no bulk write operation will be executed and this field will be unset. This field may also be unset if the bulk
+write operation is unacknowledged as permitted by the [CRUD API Spec](../crud/crud.md#write-results).
+
+See
+[Why does rewrapManyDataKey return RewrapManyDataKeyResult instead of BulkWriteResult?](#why-does-rewrapmanydatakey-return-rewrapmanydatakeyresult-instead-of-bulkwriteresult).
```

Comment:
```suggestion
[Why does rewrapManyDataKey return RewrapManyDataKeyResult instead of BulkWriteResult?](#why-does-rewrapmanydatakey-return-rewrapmanydatakeyresult-instead-of-bulkwriteresult)
```